frontend/src/components/projectEdit/partnersListing.js

```diff
@@ -0,0 +1,427 @@
+import { useEffect, useState } from 'react';
+import { useParams } from 'react-router-dom';
+
+import { useSelector } from 'react-redux';
+import ReactDatePicker from 'react-datepicker';
+import { FormattedMessage } from 'react-intl';
+import Popup from 'reactjs-popup';
+import { Tooltip } from 'react-tooltip';
+import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
+import ReactPlaceholder from 'react-placeholder';
+import { format } from 'date-fns';
+import toast from 'react-hot-toast';
+import PropTypes from 'prop-types';
+
+import messages from './messages';
+import { Alert } from '../alert';
+import { BanIcon, CircleMinusIcon, CircleExclamationIcon } from '../svgIcons';
+import { Button } from '../button';
+import { styleClasses } from '../../views/projectEdit';
+import { fetchLocalJSONAPI, pushToLocalJSONAPI } from '../../network/genericJSONRequest';
+import { DateCustomInput } from './partnersForm';
+
+const TableContentPlaceholder = () => (
+  <>
+    <tr>
+      {new Array(4).fill(
+        <td>
+          <ReactPlaceholder
+            type="rect"
+            style={{ width: '100%', height: 50 }}
+            showLoadingAnimation
+          />
+        </td>,
+      )}
+    </tr>
+    <tr>
+      {new Array(4).fill(
+        <td>
+          <ReactPlaceholder
+            type="rect"
+            style={{ width: '100%', height: 50 }}
+            showLoadingAnimation
+          />
+        </td>,
+      )}
+    </tr>
+  </>
+);
+
+export const Listing = ({ partnerIdToDetailsMapping }) => {
+  const [selectedPartner, setSelectedPartner] = useState({});
+  const [errorMessage, setErrorMessage] = useState({});
+  const [actionType, setActionType] = useState(''); // "edit" or "remove"
+
+  const token = useSelector((state) => state.auth.token);
+  const { id } = useParams();
+  const queryClient = useQueryClient();
+
+  useEffect(() => {
+    if (actionType.length === 0) {
+      setSelectedPartner({});
+    }
+  }, [actionType]);
+
+  useEffect(() => {
+    const startDate = selectedPartner.startedOn && new Date(selectedPartner.startedOn);
+    const endDate = selectedPartner.endedOn && new Date(selectedPartner.endedOn);
+
+    if (!endDate && errorMessage.id === messages.partnerEndDateError.id) {
+      setErrorMessage({});
+      return;
+    }
+
+    if (endDate && startDate < endDate && errorMessage.id === messages.partnerEndDateError.id) {
+      setErrorMessage({});
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [selectedPartner.startedOn, selectedPartner.endedOn]);
+
+  const {
```

Comment:
it would be nice to have this complex query abstracted in `api/projects.js` file